src/circuit/serde.rs

```diff
@@ -0,0 +1,216 @@
+
+    #[error("other error: {0:?}")]
+    OtherError(#[from] std::io::Error),
+}
+pub trait FromEccSerde {
+    fn deserialize_from<R: Read>(reader: R) -> Self;
+}
+
+impl<T: FromEccSerde> FromEccSerde for Vec<T> {
+    fn deserialize_from<R: Read>(mut reader: R) -> Self {
+        let vec_len = <usize as FieldSerde>::deserialize_from(&mut reader).unwrap();
+        let mut ret = vec![];
+        for _ in 0..vec_len {
+            ret.push(T::deserialize_from(&mut reader));
+        }
+        ret
+    }
+}
+
+impl<T1: FromEccSerde, T2: FromEccSerde> FromEccSerde for (T1, T2) {
+    fn deserialize_from<R: Read>(mut reader: R) -> Self {
+        (
+            T1::deserialize_from(&mut reader),
+            T2::deserialize_from(&mut reader),
+        )
+    }
+}
+
+impl FromEccSerde for usize {
+    fn deserialize_from<R: Read>(reader: R) -> Self {
+        <usize as FieldSerde>::deserialize_from(reader).unwrap()
+    }
+}
+
+impl<C: GKRConfig, const INPUT_NUM: usize> FromEccSerde for Gate<C, INPUT_NUM> {
+    fn deserialize_from<R: Read>(mut reader: R) -> Self {
+        let mut i_ids = [0usize; INPUT_NUM];
+        for id in &mut i_ids {
+            *id = <usize as FieldSerde>::deserialize_from(&mut reader).unwrap();
+        }
+
+        let o_id = <usize as FieldSerde>::deserialize_from(&mut reader).unwrap();
+
+        let coef_type_u8 = u8::deserialize_from(&mut reader).unwrap();
+        let (coef_type, coef) = match coef_type_u8 {
+            1 => (
+                CoefType::Constant,
+                C::CircuitField::try_deserialize_from_ecc_format(&mut reader).unwrap(),
+            ),
+            2 => (CoefType::Random, C::CircuitField::ZERO),
+            3 => {
+                if INPUT_NUM > 0 {
+                    panic!("Public Input can only be used with constant gates")
+                };
+
+                (
+                    CoefType::PublicInput(
+                        <usize as FieldSerde>::deserialize_from(&mut reader).unwrap(),
+                    ),
+                    C::CircuitField::ZERO,
+                )
+            }
+            _ => unreachable!(),
+        };
+
+        Self {
+            i_ids,
+            o_id,
+            coef_type,
+            coef,
+            gate_type: 0,
+        }
+    }
+}
+
+pub struct CustomGateWrapper<C: GKRConfig, const INPUT_NUM: usize> {
+    pub custom_gate: Gate<C, INPUT_NUM>,
+}
+
+impl<C: GKRConfig, const INPUT_NUM: usize> FromEccSerde for CustomGateWrapper<C, INPUT_NUM> {
+    fn deserialize_from<R: Read>(mut reader: R) -> Self {
+        let gate_type = <usize as FieldSerde>::deserialize_from(&mut reader).unwrap();
+        let i_ids: [usize; INPUT_NUM] = Vec::<usize>::deserialize_from(&mut reader)
+            .try_into()
+            .unwrap();
+
+        let o_id = <usize as FieldSerde>::deserialize_from(&mut reader).unwrap();
+
+        let coef_type_u8 = u8::deserialize_from(&mut reader).unwrap();
+        let (coef_type, coef) = match coef_type_u8 {
```

Comment:
does this start from `0` or `1`?

Reply:
It starts from 1 because that is how it's defined in ECC.